gh-134939: Add the interpreters Module (#133958)
python/cpython: Doc/whatsnew/3.14.rst

New features
============

.. _whatsnew314-pep734:

PEP 734: Multiple Interpreters in the Stdlib
--------------------------------------------

The CPython runtime supports running multiple copies of Python in the
same process simultaneously and has done so for over 20 years.
Each of these separate copies is called an "interpreter".

> However, the feature has been available only through the C-API.

Reviewer: ```suggestion
However, the feature had been available only through the C-API.
```

That limitation is removed in the 3.14 release,

> with the new :mod:`concurrent.interpreters` module!

Reviewer: ```suggestion
with the new :mod:`concurrent.interpreters` module.
```

There are at least two notable reasons why using multiple interpreters
is worth considering:

* they support a new (to Python), human-friendly concurrency model
* true multi-core parallelism

For some use cases, concurrency in software enables efficiency and
can simplify software, at a high level.  At the same time, implementing
and maintaining all but the simplest concurrency is often a struggle
for the human brain.  That especially applies to plain threads

> (e.g. :mod:`threading`), where all memory is shared between all threads.

Reviewer: Avoid Latin abbreviations: https://devguide.python.org/documentation/style-guide/#use-simple-language
```suggestion
(for example, :mod:`threading`), where all memory is shared between all threads.
```

  interpreters)
* there aren't many options *yet* for truly sharing objects or other
  data between interpreters (other than :type:`memoryview`)
* many extension modules on PyPI are not compatible with multiple
  interpreters yet (stdlib extension modules *are* compatible)
* the approach to writing applications that use multiple isolated
  interpreters is mostly unfamiliar to Python users, for now

The impact of these limitations will depend on future CPython
improvements, how interpreters are used, and what the community solves
through PyPI packages.  Depending on the use case, the limitations may
not have much impact, so try it out!

Furthermore, future CPython releases will reduce or eliminate overhead
and provide utilities that are less appropriate on PyPI.  In the
meantime, most of the limitations can also be addressed through
extension modules, meaning PyPI packages can fill any gap for 3.14, and
even back to 3.12 where interpreters were finally properly isolated and
stopped sharing the :term:`GIL`.  Likewise, we expect to slowly see
libraries on PyPI for high-level abstractions on top of interpreters.

Regarding extension modules, work is in progress to update some PyPI

> projects, as well as tools like Cython, PyBind11, Nanobind, and PyO3.

Reviewer: ```suggestion
projects, as well as tools like Cython, pybind11, nanobind, and PyO3.
```

The steps for isolating an extension module are found at
:ref:`isolating-extensions-howto`.  Isolating a module has a lot of
overlap with what is required to support

> :ref:`free-threadeding <whatsnew314-free-threaded-cpython>`,

Reviewer: ```suggestion
:ref:`free-threading <whatsnew314-free-threaded-cpython>`,
```

* they support a new (to Python), human-friendly concurrency model
* true multi-core parallelism

For some use cases, concurrency in software enables efficiency and
can simplify software, at a high level.  At the same time, implementing
and maintaining all but the simplest concurrency is often a struggle
for the human brain.  That especially applies to plain threads
(e.g. :mod:`threading`), where all memory is shared between all threads.

With multiple isolated interpreters, you can take advantage of a class
of concurrency models, like CSP or the actor model, that have found
success in other programming languages, like Smalltalk, Erlang,
Haskell, and Go.  Think of multiple interpreters like threads
but with opt-in sharing.

Regarding multi-core parallelism: as of the 3.12 release, interpreters
are now sufficiently isolated from one another to be used in parallel.
(See :pep:`684`.)  This unlocks a variety of CPU-intensive use cases
for Python that were limited by the :term:`GIL`.

Using multiple interpreters is similar in many ways to
:mod:`multiprocessing`, in that they both provide isolated logical
"processes" that can run in parallel, with no sharing by default.

> .  However, when using multiple interpreters, an application will use

Reviewer 2: ```suggestion
However, when using multiple interpreters, an application will use
```

Author: fixed

fewer system resources and will operate more efficiently (since it
stays within the same process).  Think of multiple interpreters as
having the isolation of processes with the efficiency of threads.

.. XXX Add an example or two.
.. XXX Link to the not-yet-added HOWTO doc.

While the feature has been around for decades, multiple interpreters
have not been used widely, due to low awareness and the lack of a stdlib
module.  Consequently, they currently have several notable limitations,
which will improve significantly now that the feature is finally
going mainstream.

Current limitations:

* starting each interpreter has not been optimized yet
* each interpreter uses more memory than necessary
  (we will be working next on extensive internal sharing between
  interpreters)
* there aren't many options *yet* for truly sharing objects or other
  data between interpreters (other than :type:`memoryview`)
* many extension modules on PyPI are not compatible with multiple
  interpreters yet (stdlib extension modules *are* compatible)
* the approach to writing applications that use multiple isolated

>   interpreters is mostly unfamiliar to Python users, for now

Reviewer 3: I'd recommend capitalizing sentences and add a period at the end of each sentence, e.g.:

```diff
-* starting each interpreter has not been optimized yet
+* Starting each interpreter has not been optimized yet.
```

Reviewer 4: I think the current style is fine for now

The impact of these limitations will depend on future CPython
improvements, how interpreters are used, and what the community solves
through PyPI packages.  Depending on the use case, the limitations may
not have much impact, so try it out!

Furthermore, future CPython releases will reduce or eliminate overhead
and provide utilities that are less appropriate on PyPI.  In the
meantime, most of the limitations can also be addressed through
extension modules, meaning PyPI packages can fill any gap for 3.14, and
even back to 3.12 where interpreters were finally properly isolated and
stopped sharing the :term:`GIL`.  Likewise, we expect to slowly see
libraries on PyPI for high-level abstractions on top of interpreters.

Regarding extension modules, work is in progress to update some PyPI

> projects, as well as tools like Cython, PyBind11, Nanobind, and Py03.

Reviewer 5: typo: `Py03` (zero-three) -> `PyO3`

Reviewer 4: ```suggestion
projects, as well as tools like Cython, PyBind11, Nanobind, and PyO3.
```

Author: fixed

The steps for isolating an extension module are found at
:ref:`isolating-extensions-howto`.  Isolating a module has a lot of
overlap with what is required to support
:ref:`free-threadeding <whatsnew314-free-threaded-cpython>`,
so the ongoing work in the community in that area will help accelerate
support for multiple interpreters.

Also added in 3.14: :ref:`concurrent.futures.InterpreterPoolExecutor

> <whatsnew314-concurrent-futures-interp-pool>`.

Reviewer 3: I'd recommend adding a ".. seealso: :pep:`XXX`" at the end of the section as for other PEPs

Author: fixed
